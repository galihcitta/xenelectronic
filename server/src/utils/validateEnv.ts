import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
