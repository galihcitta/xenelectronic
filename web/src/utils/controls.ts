import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
