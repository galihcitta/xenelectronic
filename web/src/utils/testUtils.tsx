import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
