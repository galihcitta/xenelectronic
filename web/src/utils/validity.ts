import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
