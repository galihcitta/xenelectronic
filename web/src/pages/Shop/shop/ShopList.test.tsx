import { render, fireEvent } from '@testing-library/react';
import { withProvider } from '../../../utils/testUtils';
import { _ShopList, ShopListProps } from './ShopList.component';

const ShopListWithProvider = withProvider(_ShopList);

describe('<ShopList />', () => {
  const mockFetchProductsByType = jest.fn();
  const mockIncreaseLimit = jest.fn();

  const props: ShopListProps = {
    filteredProducts: [
      {
        character: 'modern',
        collection: 'clocks',
        color: 'black',
        description: 'lorem ipsum',
        id: '1',
        images: ['aa', 'bb'],
        name: 'Clock',
        price: '200',
        type: 'clocks'
      },
      {
        character: 'modern',
        collection: 'clocks',
        color: 'black',
        description: 'lorem ipsum',
        id: '2',
        images: ['aa', 'bb'],
        name: 'Clock',
        price: '200',
        type: 'clocks'
      },
      {
        character: 'modern',
        collection: 'clocks',
        color: 'black',
        description: 'lorem ipsum',
        id: '3',
        images: ['aa', 'bb'],
        name: 'Clock',
        price: '200',
        type: 'clocks'
      }
    ],
    filteredProductsAmount: 3,
    productsFetched: true,
    limit: 2,
    query: '',
    gridColumns: 2,
    fetchProductsByType: mockFetchProductsByType,
    increaseLimit: mockIncreaseLimit
  };

  const setup = (props: ShopListProps) => {
    const utils = render(ShopListWithProvider(props));
    const productsList = utils.container.querySelector('ul');

    return { ...utils, productsList };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('products list', () => {
    it('contains correct style rule', () => {
      const { productsList } = setup(props);
      expect(productsList).toHaveStyle('grid-template-columns: repeat(2, 1fr)');
    });
  });

  it('calls `increaseLimit` action', () => {
    const { getByText, rerender, container } = setup(props);

    fireEvent.click(getByText(/more results/i));
    expect(mockIncreaseLimit).toHaveBeenCalledTimes(1);

    rerender(ShopListWithProvider({ ...props, limit: 4 }));
    expect(container.querySelector('button[type="button"]')).toBeNull();
    expect(container.querySelectorAll('li').length).toBe(3);
  });

  it('contains warning with text `no results`', () => {
    const { getByText } = setup({ ...props, filteredProducts: [] });
    expect(getByText(/no results/i)).toBeInTheDocument();
  });

  it('calls `fetchProductsByType` action', () => {
    const { rerender } = setup(props);
    rerender(ShopListWithProvider({ ...props, productsFetched: false }));

    expect(mockFetchProductsByType).toBeCalledWith('all');
  });
});
