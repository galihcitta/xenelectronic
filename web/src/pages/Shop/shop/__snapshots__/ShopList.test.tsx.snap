// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShopList /> renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="sc-bdVaJa eOVwsA"
  >
    <a
      href="/product/1"
    >
      <li
        class="sc-ifAKCX gAuHj"
      >
        <figure
          class="sc-htpNat iSLqsH"
        >
          <img
            alt="Clock"
            class="sc-bxivhb lelnaL"
            src="aa"
          />
        </figure>
        <div
          class="sc-EHOje gmmqIO"
        >
          <p
            class="sc-bZQynM fGwzMD"
          >
            Clock
          </p>
          <p
            class="sc-gzVnrw hUlLlY"
          >
            $ 200
          </p>
        </div>
      </li>
    </a>
    <a
      href="/product/2"
    >
      <li
        class="sc-ifAKCX gAuHj"
      >
        <figure
          class="sc-htpNat iSLqsH"
        >
          <img
            alt="Clock"
            class="sc-bxivhb lelnaL"
            src="aa"
          />
        </figure>
        <div
          class="sc-EHOje gmmqIO"
        >
          <p
            class="sc-bZQynM fGwzMD"
          >
            Clock
          </p>
          <p
            class="sc-gzVnrw hUlLlY"
          >
            $ 200
          </p>
        </div>
      </li>
    </a>
  </ul>
  <button
    class="sc-htoDjs dBUMDP"
    type="button"
  >
    More results (1)
  </button>
</DocumentFragment>
`;
